### PR TITLE
Block API: Support nesting (InnerBlocks) in unknown block types

### LINKED ISSUE #13391: Inner Blocks are lost if Parent Container Block is removed.
There are many plugins available that create a Container ( Section or Row ) block that allow Inner Blocks to be added. Examples: Atomic Blocks, CoBlocks, Ultimate Addons for Gutenberg etc. Deactivating any of the plugins tested results in the following warning within the Editor.

screenshot 2019-01-20 at 18 47 58

If the user chooses to **Keep as HTML** or switches to the Code Editor then all of the Inner Block content is lost. There is no ability to extract the Inner Blocks as a Common Block or as HTML.

**Replicate this issue**

NOTE: I have chosen the Atomic Block plugin for this example, but as mentioned it effects all of these types of plugin.

WordPress 5.0.3
Gutenberg 4.8
Atomic Blocks  1.4.23

1. Install Atomic Blocks plugin from WordPress repository
2. Add a AB Container Block.
3. Populate the AB Container Block with a single Title Block.
4. Switch to Code Editor and this is the resulting output:

screenshot 2019-01-20 at 19 19 29

5. Deactivate Atomic Blocks Plugin.
6. Check Page Preview to see that Inner Block Content remains intact.
7. Within Editor switch to Code Editor and this is the resulting output:

screenshot 2019-01-20 at 19 19 40

This ultimately leads to 'Block Lock'. The only way the user can retain the content added to a Container Block is to extract them before the plugin is deleted. This could lead to loss of user content within the Editor. 

Proposed solution: Option to extract all blocks as separate HTML or Common Blocks

### LINKED ISSUE #15202: Inner blocks of a missing parent are lost on save
**Describe the bug**
When a post contains a missing (i.e., unregistered) parent block containing inner blocks, its inner blocks are omitted from the content when the post is next saved.

**To reproduce**
1. Open the block editor to create a new post.
2. Add a columns block with a paragraph in each column.
3. Safe the post as a draft.
4. Open the same post in the Classic Editor and note the presence of the `wp:columns` block and its inner blocks.
5. Rename the opening and closing `<!-- wp:columns -->` delimiters to use the name `wp:columns-missing` and save the draft.
6. Reload the Classic Editor and note the presence of the `wp:columns-missing` block and its inner blocks.
6. Open the post in the block editor and note the missing block notice.
7. Add a paragraph block and save again.
8. Open the post in Classic Editor and note the presence of `wp:columns-missing` block and the absence of inner blocks.

**Expected behavior**
The inner blocks of the `wp:columns-missing` block would be preserved in the post content.

### PR DESCRIPTION
**Pull request by mcsf**
Fixes #13391.
Fixes #15202.

## Description

When the block parser finds a block of unknown type, it wraps it in a special block (per `getUnregisteredTypeHandlerName`) — by default, this will be a block of type `core/missing`. This new outer block can then offer the user the possibility to extract the wrapped content.

This pull request updates the parsing logic so that the fallback block is fed the entire tree of content of the unknown block — if it has child blocks — and not just its own surface-level content.

## Example

When parsing the following unknown block:

```html
<!-- wp:my/unknown -->
  <p>Begin block that contains a list.</p>
  <!-- wp:list -->
  <ul><li>1</li><li>2</li></ul>
  <!-- /wp:list -->
  <p>End a block that contains a list.</p>
<!-- /wp:my/unknown -->
```

Before, the rescued `originalContent` would have been:

```html
<!-- wp:my/unknown -->
  <p>Begin block that contains a list.</p>
  <p>End block that contains a list.</p>
<!-- /wp:my/unknown -->
```

Now, it would be the entire markup for `my/unknown`.

## How has this been tested?

See parent issue for details. In addition, below are some concrete steps:

1. Start a new post.
1. Add a Columns block with two columns.
1. Add some basic content to each column.
1. Back in the source code, comment out the registration of the Columns block (see diff below).
1. Refresh the editor.

- The former Columns block should be singled out as an unknown block.
- Its in-editor preview should still reveal the former content, though greyed out.
- Actioning _Keep as HTML_ should replace it with an HTML block reflecting the original content.
- Not actioning _Keep as HTML_ and previewing the whole post should reflect the original content.
- Switching to the Code Editor should reveal the whole nested content.

```diff
diff --git a/packages/block-library/src/index.js b/packages/block-library/src/index.js
index 81a0e9bec..6a7c4569c 100644
--- a/packages/block-library/src/index.js
+++ b/packages/block-library/src/index.js
@@ -86,7 +86,7 @@ export const registerCoreBlocks = () => {
 		calendar,
 		categories,
 		code,
-		columns,
+		// columns,
 		column,
 		cover,
 		embed,
```

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by alvaro0555** [comment]
Thanks for the effort to fix this. How does it work when there are nested InnerBlocks (in the test, commenting out also `column`)?

**Comment by mcsf** [comment]
> Thanks for the effort to fix this. How does it work when there are nested unknown blocks with InnerBlocks (in the test, commenting out also `column`)?

@alvaro0555: It should work all the same (I re-tested by disabling both `core/columns` and `core/column`). This is because the parser stops* at the outermost unknown block and blindly reserialises its inner blocks to produce the `core/missing` block.

*: After which the parser continues parsing laterally.

**Comment by diggeddy** [comment]
@mcsf thank you, really appreciate the effort.

**Comment by diggeddy** [comment]
@dmsnell any update on if/when this will be merged?

**Comment by youknowriad** [comment]
Something we need to take over as @mcsf will be AFK for a few weeks @WordPress/gutenberg-core

**Comment by dmsnell** [comment]
@diggeddy I away from the office for about two weeks, so I apologize for my delay. I will try and spend some time on this before the weekend. If not, please ping me again and if I've been unable to address it I'll likely approve it and merge to prevent it being a roadblock.

**Comment by diggeddy** [comment]
@dmsnell any news ? be good to see this put to bed.

**Comment by dmsnell** [comment]
Thanks for the extra ping @diggeddy.

I would like feedback on what I write here, as I'm a bit skeptical still of the `reconstitute`-approach for solving the problem.

I'm not sure what we are trying to gain by reserializing content inside the fallback block. **Do we not already have the full tree of `innerBlocks` and `innerContent` when we create the fallback block? Well, I think we _need_ to pass along `innerContent` and `innerBlocks` (note that `innerHTML` should not be used anymore. we need to add some official deprecation warning) but as for reserialization I'm confused.

**Are we wanting to extract content inside a container as one big blob of potentially-block-containing HTML?** Why not provide an option to extract inner blocks _as blocks_ instead?

So my thinking is that if we pass in `innerContent` and `innerBlocks` to the fallback then we can tell it to extract the insides and at that time insert new blocks. We shouldn't need to print out and then re-parse or even consider that as a separate _kind_ of thing in these cases.

Not sure if @mcsf is back yet or not. If not, I can update this diff with my suggested changes, or provide another PR for it. I'd like to get a picture of how we want to handle these changes in the UI since this PR seems pretty limited to providing information that's currently missing. Is there already another PR to provide the author-facing functionality?

```js
const blocks = [];
let blockIndex = 0;
for ( let chunk of innerContent ) {
  if ( null === chunk ) {
    blocks.push( innerBlocks[ blockIndex++ ] );
  } else {
    blocks.push( paragraphBlockFromHTML( chunk ) );
  }
}
```

With this approach we wouldn't need `reconstituteBlockNode` and we wouldn't have to make a distinction between `delimited` and `undelimited`.

As I don't see the updates to the fallback block which provide the _extract_ functionality I'm thinking this may be more of an implementation question than anything else. As I see it though from the perspective of parsing producing a tree it feels more natural to me to extract _blocks_ instead of _html_. It answers the recursive problem because if we have an inner block we also don't understand we can further extract _it_ too if we want, but in a separate click.

**Comment by tomusborne** [comment]
From the perspective of a user, being able to extract the inner blocks makes more sense than being left with a bunch of HTML.

Maybe it should be an option though? Something like a "Convert to Blocks" button? That will allow them to choose whether to convert the content to blocks or re-activate the plugin with the container.

**Comment by diggeddy** [comment]
@dmsnell I agree with @tomusborne on this... any chance we may see this PR as a near future option?

**Comment by tomusborne** [comment]
The button I mentioned likely isn't even necessary if it complicates things, the user could just leave the page and not save if they decide to re-activate the plugin with the container block.

**Comment by diggeddy** [comment]
@dmsnell any news on this? Sorry for the directness.

**Comment by dmsnell** [comment]
@diggeddy your patience is exemplary and I'm grateful for your pings - don't ever feel bad about asking me for updates (I'm generally working on several important things at once).

To that end I created #15078 to create a working model for the solution I proposed, which was to extract blocks.

What I'm finding harder to get behind with this PR's solution is creating new concepts and special cases in the parsed block object. I believe that we can accomplish what @mcsf proposed not only by _not_ introducing new things into the system but by _removing_ special cases.

Therefore I'm proposing #15078 as a tiny incremental improvement on the behavior which I hope meets your immediate needs while offering some addition time to fix the broader issue with unsupported blocks holding inner blocks. A parser change may help out but we may not need it or need as much as we think we do.

**Comment by mcsf** [comment]
I'm back and able to follow up on this.

> As I don't see the updates to the fallback block which provide the _extract_ functionality I'm thinking this may be more of an implementation question than anything else. […]

No other updates are needed in order for this to allow the rescuing of content nested in an unknown block, which was the object of the parent issue.

Further, in addressing that, `core/missing` now has the actual original data, as found in the post data, from which it could trivially build its own block tree if it so wanted (e.g. to offer the _Extract inner blocks_ functionality mentioned in previous comments):

```js
// in core/missing
const tree = wp.blockSerializationDefaultParser.parse( attributes.originalContent );
const branch = tree[ 0 ].innerBlocks;
```

This benefit extends to any non-core blocks that one might want to register as `unregisteredTypeHandlerName`.

Compared with the alternative #15078 opened by @dmsnell, this solution comes with a smaller surface area. In particular, the interface between the blocks parser and blocks themselves is kept intact; we just ensure that `originalContent` and `originalUndelimitedContent` make sense when dealing with nested content.

My personal impression is that a good deal is easily taken for granted in this PR. A cursory review of #15078 confirms this, as concerns the API (attribute schema) changes of `core/missing`, and more worryingly the legwork needed around `save()`. Depending on dedicated logic to ensure we never lose users' content upon saving is something I can't say I'm comfortable with. Meanwhile, in only dealing with serialisation-ready attributes, the present PR automatically fixes saving; for this reason I've added `Fixes #15202.` to the description.

> I would like feedback on what I write here, as I'm a bit skeptical still of the `reconstitute`-approach for solving the problem.

I understand that the reconstitution attracts attention and that any developer would wonder about alternatives. But I don't think I've seen concrete arguments on why one should be favoured over another, especially if we look at the ultimate goal of ensuring nesting support with no bugs or other hiccups.

To sum up a little:

- Since the release of WordPress 5.0 more than ever, we should avoid changing APIs, especially in core blocks, unless there is a compelling reason for it.
- Data safety is paramount, and I feel more confident about fake-reconstituting a block node at parse than fake-serialising a block at save.

**Comment by mcsf** [comment]
I've pushed 87c47c3, which adds a secondary action to keep an unrecognised block's inner blocks, if any are present.

extract-inner-blocks

This can be tested in four scenarios:

1. With a simple unknown block containing no children, the only possible action should be _Keep as HTML_. Behaviour unchanged by this pull request.
2. With an unknown block containing generic children, the two possible actions should be _Keep as HTML_ and _Keep inner blocks_. Pressing _Keep as HTML_ should replace it with an HTML block containing the actual deep HTML. Pressing _Keep inner blocks_ should replace it with the expected inner blocks.
3. With an unknown block containing children which **depend on the former** (e.g. `core/column` states `parent: [ 'core/columns' ]`) and are **also unknown**, pressing _Keep inner blocks_ should replace it with other unknown blocks corresponding to the unknown children.
4. With an unknown block containing children which **depend on the former** (e.g. `core/column` states `parent: [ 'core/columns' ]`) but are **known**, pressing _Keep inner blocks_ should yield no action. This is because the children aren't allowed to exist outside of the unknown parent, and thus the `replaceBlock` action **fails**. One workaround would be to keep traversing the children until no "dependent" blocks were found, but given its brittleness this seems like an edge scenario **not worth handling**.

**Comment by mcsf** [comment]
> I've pushed [87c47c3](https://github.com/WordPress/gutenberg/commit/87c47c3d663f6fa1c5c6468285b7642fe74e9c0e), which adds a secondary action to keep an unrecognised block's inner blocks, if any are present.

Note that I would personally leave this commit out of the pull request, to keep it focused on the bug fix. 87c47c3 should be cherry-picked for an enhancement PR.

**Comment by Tug** [comment]
I agree we should leave https://github.com/WordPress/gutenberg/commit/87c47c3d663f6fa1c5c6468285b7642fe74e9c0e out of this PR. Although it's a nice addition it might need more work. For instance we could check if all blocks can be inserted first and display a message or hide the button if it's not the case.

Reverting for now

**Review by Tug — APPROVED** [review]
This works for me and also fixes our issue on native https://github.com/wordpress-mobile/gutenberg-mobile/issues/1206

I agree we should deprecate `block.innerHTML` as it seems redundant with `innerContent`.
Another solution would be to update it to return the real inner HTML just like with `Element` on the dom. But blocks being pure objects I don't think it would be a good idea to throw a getter here.
Anyway, this should be done in a follow up PR.

File: packages/blocks/src/api/test/parser.js [hunk 1]
```diff
@@ -757,6 +758,111 @@ describe( 'block parser', () => {
...
+							<!-- /wp:list -->
+							<!-- wp:paragraph -->
+							<p>D</p>
+							<!-- /wp:paragraph -->
+						</div></div>
+						<!-- /wp:group -->
+					</div>
+					<!-- /wp:column -->
+				</div>
+				<!-- /wp:columns -->`.replace( /\t/g, '' );
```
Comment (mcsf) [inline packages/blocks/src/api/test/parser.js]: <3 thanks for cleaning up.

**Review by mcsf — COMMENTED** [review]
Thanks for the help, @Tug!

File: packages/blocks/src/api/parser.js [hunk 2]
```diff
@@ -416,17 +417,39 @@ export function createBlockWithFallback( blockNode ) {
 	let blockType = getBlockType( name );
 
 	if ( ! blockType ) {
-		// Preserve undelimited content for use by the unregistered type handler.
-		const originalUndelimitedContent = innerHTML;
+		// Since the constituents of the block node are extracted at the start
```
Comment (mtias) [inline packages/blocks/src/api/parser.js]: Thanks for all the comments.
